core/src/main/java/org/mule/routing/AsynchronousUntilSuccessfulProcessingStrategy.java

```diff
@@ -215,8 +217,11 @@ public static Serializable buildQueueKey(final MuleEvent muleEvent)
         // the key is built in way to prevent UntilSuccessful workers across a
         // cluster to compete for the same
         // events over a shared object store
-        String key = muleEvent.getFlowConstruct() + "-" + muleEvent.getMuleContext().getClusterId() + "-"
-                     + muleEvent.getId();
+        // it also adds a random trailer to support events which have been
+        // splitted and thus have the same id. Random number was chosen over
+        // UUID for performance reasons
+        String key = String.format("%s-%s-%s-%d", muleEvent.getFlowConstruct(), muleEvent.getMuleContext().getClusterId(), muleEvent.getId(), random.nextInt());
```

Comment:
No problem with making the key random. Despite it exceeds the purpose of this PR, looking at the reason to build the message key that way, I see an existing issue when a server is shutdown gracefully or crashes.

Currently each node schedules for processing the events that it receives. So every node in a cluster would process the message in which the until-successful was executed. 
If a new node is added to the cluster then it would execute the start phase of the until-successful component which would schedule EVERY message in the object store to be process in that new node which may lead to race conditions between the new node and the node that store the event the first time.

Also it doesn't seem to manage well failover. When a node goes done all the pending tasks of the until-successful will not be executed until the node gets restarted.